### PR TITLE
Added ChibiOS RTOS Global Variable, Virtual Timers and Statistics, reworked createHmlHelp and refresh

### PR DESCRIPTION
**Pull request by vrepetenko**

### CODE REVIEW

**Comment by haneefdm** [comment]
It is not clear to me if this was handled but we expect more than one RTOS at the same time. This happens in multu-core debug where each core may be running an RTOS. So, the state info may have to be session specific. I would use the session name.

**Comment by vrepetenko** [comment]
> It is not clear to me if this was handled but we expect more than one RTOS at the same time. This happens in multu-core debug where each core may be running an RTOS. So, the state info may have to be session specific. I would use the session name.

I think you mean this part:

```
            for (const rtosSession of this.sessionMap.values()) {
                promises.push(rtosSession.updateUIElementState(elementId, state));
            }
```

    should be:

   ` promises.push(currentSession.updateUIElementState(elementId, state));`
    
    But where currentSession can be found?

**Comment by haneefdm** [comment]
There is no concept of the current session. We need to pass the session-name to the new and all settings can be then stored as a combination of `<session-name>-<rtos-name>-propname`. So a chance would be specific to a session and its rtos.

I have to look in the code to figure out how to pass the session-name to the webview side

**Comment by vrepetenko** [comment]
> We can make each RTOS session info under a div whose element-id would be the session-name. Session names will be unique as VSCode will not allow dups (AFAIK).
> 
> So, if you extract the full path to your element, then that could be the key.

Do you mean full Dom path?
Can not get what problem we are trying to solve with this path...
Now any RTOS implementation can use 'rtos-panels' or may not use it at all.
Or xRTOS extension uses the same sessions list in different VSCODE windows, so if you open two VSCODE windows and run two different debug session there will be a problem with rtos-panels change event processing?

**Comment by haneefdm** [comment]
Okay, not sure we are on the same wavelength here. In a multi-core setup, we can have more than one debug session going at the same time. One session can use FreeRTOS and another can use ChibiOS.  We put in a whole lot of effort to make this happen in a single instance of VSCode.

https://github.com/Marus/cortex-debug/wiki/Multi-core-debugging

You can also have multiple sessions independent of the cortex-debug method in the same instance of VSCode.

And, I am not talking about two different VSCode windows. That is a completely different topic.

**Comment by vrepetenko** [comment]
it is clear now, thanks for explanation!
I will try to find a solution...

**Comment by haneefdm** [comment]
Hmmm. It appears that a build has failed. Looks like `npm run compile` failed. Something wrong with package.json or node_modules?

**Review by haneefdm — APPROVED** [review]

**Comment by haneefdm** [comment]
@PhilippHaefele when you get a chance, could you review this PR?

File: tsconfig.json [hunk 1]
```diff
@@ -12,6 +12,7 @@
         "sourceMap": true,
         "allowJs": true,
         "checkJs": false,
+        "skipLibCheck": true,
```
Comment (PhilippHaefele) [inline tsconfig.json:15]: Nice one, didn't know that one

File: src/rtos/rtos-chibios.ts [hunk 2]
```diff
@@ -41,56 +43,136 @@ enum chThreadState {
...
+                           { columnDataKey: 'last', title: 'Last Deadline' },
+                           { columnDataKey: 'reload', title: 'Reload' }];
+
+const statisticsCols = [{ columnDataKey: 'description', title: 'Measured Section' },
+                        { columnDataKey: 'best', title: 'Best Case' },
+                        { columnDataKey: 'worst', title: 'Worst Case' },
+                        { columnDataKey: 'counter', title: 'Iterations' },
+                        { columnDataKey: 'cumulative', title: 'Cumulative Time' } ];
+
+const traceCols = [{ columnDataKey: 'event', title: 'Event' },
```
Comment (PhilippHaefele) [inline src/rtos/rtos-chibios.ts:133]: Is it intended that `traceCols` is not used?

File: src/rtos/rtos-chibios.ts [hunk 3]
```diff
@@ -596,16 +959,58 @@ export class RTOSChibiOS extends RTOSCommon.RTOSBase {
...
+                                                           [{ name: 'id', value: 'timers' },
+                                                            { name: 'aria-label', value: 'Virtual Timers' }]);
+
+        const htmlStatistics = this.getHTMLDataGrid(statisticsCols,
+                                                    this.statistics,
+                                                    [{ name: 'id', value: 'statistics' },
+                                                     { name: 'aria-label', value: 'Statistics' }]);
+
+        const htmlRTOSPanels = this.getHTMLPanels([{ title: 'GLOBAL' },
+                                                   { title: `THREADS
```
Comment (PhilippHaefele) [inline src/rtos/rtos-chibios.ts:989]: I would like to see the threads as the default (first) entry.

I think this is the first thing a user wants to see in the first place to get an overview of the OS state.

@haneefdm Any opinion regarding this?

File: src/rtos/rtos-chibios.ts [hunk 4]
```diff
@@ -1,11 +1,13 @@
 /* eslint-disable @typescript-eslint/naming-convention */
+import { breadcrumbTemplate } from '@microsoft/fast-foundation';
+import { buffer } from 'stream/consumers';
```
Comment (PhilippHaefele) [inline src/rtos/rtos-chibios.ts:3]: `buffer ` and `breadcrumbTemplate` do not seem to be used or do i overse something?

**Review by PhilippHaefele — CHANGES_REQUESTED** [review]

File: src/rtos/rtos-chibios.ts (hunk 2 above)
Comment (vrepetenko) [reply src/rtos/rtos-chibios.ts:133]: yes, trace code is not in use now. It takes to long to get trace data.

File: src/rtos/rtos-chibios.ts (hunk 4 above)
Comment (vrepetenko) [reply src/rtos/rtos-chibios.ts]: removed

File: src/rtos/rtos-chibios.ts (hunk 2 above)
Comment (PhilippHaefele) [reply src/rtos/rtos-chibios.ts:131]: 👍

**Comment by haneefdm** [comment]
> I would like to see the threads as the default (first) entry.

Agree. This would be the expectation

**Comment by haneefdm** [comment]
I would like to make a release soon. Quite a few things are batched up. I will merge this soon. I know @PhilippHaefele is busy for a couple of days but I would like him to finish his review.

Can you all update the CHANGELOG as needed if not already done?

**Comment by PhilippHaefele** [comment]
Only the default view discussion is open and after that being resolved, we're good to go from my side 👍

**Comment by vrepetenko** [comment]
> Only the default view discussion is open and after that being resolved, we're good to go from my side 👍
check this out:
https://github.com/mcu-debug/rtos-views/pull/34/commits/1bde86c7d1f42d245a82a857a03821540b9a9454

**Comment by PhilippHaefele** [comment]
Ups that somehow got hidden from me in the GutHub app 😄

**Review by PhilippHaefele — APPROVED** [review]

**Comment by vrepetenko** [comment]
> I would like to make a release soon. Quite a few things are batched up. I will merge this soon. I know @PhilippHaefele is busy for a couple of days but I would like him to finish his review.
> 
> Can you all update the CHANGELOG as needed if not already done?

@haneefdm, should I update CHANGELOG with new PR?

**Comment by haneefdm** [comment]
I already made a release over the weekend. But, sure you can edit the CHANGELOG and update what I said. Add the PR# perhaps.

It is best to update the Changelog as we submit PRs.

**Pull request merged**